Sources/Common/DataModel/Polygon/index.js

```diff
@@ -174,45 +193,629 @@ export function getBounds(poly, points, bounds) {
 
 // ---------------------------------------------------
 /**
- * Compute the normal of a polygon and return its norm.
+ * Compute the normal of a polygon.
  *
- * TBD: This does the same thing as vtkPolygon.computeNormal,
- * but in a more generic way. Maybe we can keep the public
- * static method somehow and have the private method use it instead.
- *
- * @param {Array<Number>|TypedArray<Number>} poly
+ * @param {Vector3[]} poly
  * @param {vtkPoints} points
  * @param {Vector3} normal
  * @returns {Number}
  */
 export function getNormal(poly, points, normal) {
-  normal.length = 3;
-  normal[0] = 0.0;
-  normal[1] = 0.0;
-  normal[2] = 0.0;
-
+  const n = poly.length;
+  const nn = [0.0, 0.0, 0.0];
   const p0 = [];
   let p1 = [];
   let p2 = [];
-  const v1 = [];
-  const v2 = [];
 
   points.getPoint(poly[0], p0);
   points.getPoint(poly[1], p1);
 
-  for (let j = 2; j < poly.length; j++) {
+  const v1 = [];
+  const v2 = [];
+  for (let j = 2; j < n; j++) {
     points.getPoint(poly[j], p2);
     vtkMath.subtract(p2, p1, v1);
     vtkMath.subtract(p0, p1, v2);
 
-    const n = [0, 0, 0];
-    vtkMath.cross(v1, v2, n);
-    vtkMath.add(normal, n, normal);
+    nn[0] += v1[1] * v2[2] - v1[2] * v2[1];
+    nn[1] += v1[2] * v2[0] - v1[0] * v2[2];
+    nn[2] += v1[0] * v2[1] - v1[1] * v2[0];
 
     [p1, p2] = [p2, p1];
   }
 
-  return vtkMath.normalize(normal);
+  const norm2 = vtkMath.dot(nn, nn);
+  if (norm2 > 0) {
+    const norm = Math.sqrt(norm2);
+    normal[0] = nn[0] / norm;
+    normal[1] = nn[1] / norm;
+    normal[2] = nn[2] / norm;
+  }
+
+  return norm2;
```

Comment:
Here the norm is already computed anyway, so `norm` can be returned. I think checking if `norm2` is positive does not make sense since this could not happen with real vectors and vectors having values such as `NaN` or `undefined` or `null` would anyway throw while computing the dot product.